examples/caustics.dx

```diff
@@ -0,0 +1,475 @@
+--    f i
+--  ()
+
+-- Instead of rasterizing every pixel, we can rasterize only those touched by
+-- the line segment.
+def rasterLineSegment (height:Int)?-> (width:Int)?-> (ref:Type)?->
+      (_: VSpace out)?=>
+      (a: RelPoint) -- First line endpoint
+      (b: RelPoint) -- Second line endpoint
+      (fill: out) -- Fill intensity per relative unit.
+      (canvas:Ref ref ((Fin width & Fin height)=>out)) -- Where to draw to.
+      :{Accum ref} Unit =
+  (x1, y1) = a
+  (x2, y2) = b
+  dx = 1.0 / IToF (width - 1)
+  dy = 1.0 / IToF (height - 1)
+
+  x_changes_more = abs (x1 - x2) > abs (y1 - y2)
+  (u1, u2, v1, v2, du, dv) = case x_changes_more of
+    True -> case x1 < x2 of
+      True -> (x1, x2, y1, y2, dx, dy)
+      False -> (x2, x1, y2, y1, dx, dy)
+    False -> case y1 < y2 of
+      True -> (y1, y2, x1, x2, dy, dx)
+      False -> (y2, y1, x2, x1, dy, dx)
+
+  min_u_int = FToI $ floor (u1 / du)
+  max_u_int = FToI $ ceil (u2 / du)
+  delta_u = max_u_int - min_u_int + 1
+  for u_offset:(Fin delta_u).
+    u_int = ordinal u_offset + min_u_int
+    u_rel_lo = IToF u_int * dx
+    v_rel_earlier = v1 + (u_rel_lo - u1)*(v2 - v1)/(u2 - u1)
+    v_int = FToI $ floor (v_rel_earlier / dv)
+    v_int_next = case v2 > v1 of
+      True -> v_int + 1
+      False -> v_int - 1
+    case x_changes_more of
+      True ->
+        rasterLineSegmentInFourPixels u_int v_int a b fill canvas
+        rasterLineSegmentInFourPixels u_int v_int_next a b fill canvas
+      False ->
+        rasterLineSegmentInFourPixels v_int u_int a b fill canvas
+        rasterLineSegmentInFourPixels v_int_next u_int a b fill canvas
+  ()
+
+
+:plotmat
+  s = 100
+  drawIntoCanvas s s $ \canvas.
+    rasterLineSegment (0.1, 0.3) (0.8, 0.6) 1.0 canvas
+    rasterLineSegment (0.6, 0.8) (0.3, 0.1) 1.0 canvas
+
+-- Bilinear approximation preserves line integral
+:p sum $ sum $ drawIntoCanvas 100 100 $
+  rasterLineSegment (0.1, 0.3) (0.8, 0.6) 1.0
+
+'## Light path visualization
+We can visualize light by rasterizing many lines coming from a single point.
+
+def invertAndGammaCorrect (raw : w=>h=>Float) : w=>h=>Float =
+  -for i j. exp $ 0.45 * log raw.i.j
+
+:plotmat
+  s = 100
+  n = 10000
+  invertAndGammaCorrect $ drawIntoCanvas s s $ \canvas.
+    for i:(Fin n).
+      key = ixkey (newKey 42) i
+      angle = 2. * pi * rand key
+      rasterLineSegment (0.5, 0.5) (0.5 + (cos angle), 0.5 + (sin angle)) 1.0 canvas
+    ()
+
+'But it's not very interesting unless the light has something to interact with!
+
+data BoundaryCurve =
+  LineSegment start:RelPoint end:RelPoint
+  ArcSegment center:RelPoint radius:Float theta1:Float theta2:Float  -- smallest theta must be between [-pi, pi)
+
+Ray = { start: RelPoint & dir: RelPoint }
+
+
+def invert2d (m: Fin 2 => Fin 2 => Float) : Fin 2 => Fin 2 => Float =
+  det = m.(0@_).(0@_) * m.(1@_).(1@_) - m.(0@_).(1@_) * m.(1@_).(0@_)
+  (1.0/det) .* [ [m.(1@_).(1@_), -m.(0@_).(1@_)]
+  , [-m.(1@_).(0@_), m.(0@_).(0@_)]
+  ]
```

Comment:
Maybe we should just support table unpacking syntax? You could do
```py
[[a, b], [c, d]] = m
```
and then do math on the individual elements

Reply:
That's a great idea! Yeah, that would clean up a lot of stuff.